src/osd/PrimaryLogPG.cc

```diff
@@ -10078,6 +10078,76 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
+struct C_gather : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  PrimaryLogPG::OpContext *ctx;
+  C_gather(PrimaryLogPG *pg_, hobject_t oid_, PrimaryLogPG::OpContext *ctx_) : pg(pg_), oid(oid_), ctx(ctx_) {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    pg->lock();
+    map<hobject_t,PrimaryLogPG::CLSGatherOp>::iterator p = pg->cls_gather_ops.find(oid);
+    if (p == pg->cls_gather_ops.end()) {
+      // op was cancelled
+      pg->unlock();
+      return;
+    }
+    pg->cls_gather_set_result(oid, r);
+    pg->execute_ctx(ctx);
+    pg->unlock();
+  }
+};
+
+int PrimaryLogPG::start_cls_gather(OpContext *ctx, std::shared_ptr<std::map<std::string, bufferlist> > src_obj_buffs, const std::string& pool,
+				   const char *cls, const char *method, bufferlist& inbl)
+{
+  OpRequestRef op = ctx->op;
+  MOSDOp *m = static_cast<MOSDOp*>(op->get_nonconst_req());
+  object_locator_t oloc = m->get_object_locator();
+
+  ObjectState& obs = ctx->new_obs;
+  object_info_t& oi = obs.oi;
+  const hobject_t& soid = oi.soid;
+
+  ObjectContextRef obc = get_object_context(soid, false);
+  C_GatherBuilder gather(cct);
+
+  auto cgop = CLSGatherOp(ctx, obc, op);
```

Comment:
auto &cgop = cls_gather_ops[soid];

Reply:
You should be able to avoid the second copy.

Reply:
@athanatos 

I tried the followings and somehow got compilation errors in both cases.

auto [iter, inserted] = cls_gather_ops.emplace(soid, {ctx, obc, op});
auto [iter, inserted] = cls_gather_ops.emplace({soid, {ctx, obc, op}});

Instead, I am calling the constructor of CLSGatherOp explicitly.